kasa/device_helpers.py

```diff
@@ -0,0 +1,89 @@
+"""Device creation by type."""
+
+from typing import Any, Dict, Optional, Tuple, Type
+
+from .aestransport import AesTransport
+from .connectionparams import ConnectionParameters
+from .credentials import Credentials
+from .exceptions import UnsupportedDeviceException
+from .iotprotocol import IotProtocol
+from .klaptransport import KlapTransport, TPlinkKlapTransportV2
+from .protocol import BaseTransport, TPLinkProtocol, TPLinkSmartHomeProtocol
+from .smartbulb import SmartBulb
+from .smartdevice import SmartDevice, SmartDeviceException
+from .smartdimmer import SmartDimmer
+from .smartlightstrip import SmartLightStrip
+from .smartplug import SmartPlug
+from .smartprotocol import SmartProtocol
+from .smartstrip import SmartStrip
+from .tapo import TapoBulb, TapoPlug
+
+
+def get_protocol_from_connection_parameters(
+    connection_params: ConnectionParameters,
+    host: str,
+    credentials: Optional[Credentials] = None,
```

Comment:
I would rather like to see this inside the `ConnectionParameters` class than as a separate function.

Basically, how I imagine it in my head is that it will contain *everything* necessary  to form a connection to a device. This should include the host, the credentials, and potentially any other relevant connection parameters (that is, maybe things like timeouts etc.).

This would be by no means the way you *need* to construct an instance (i.e., `SmartBulb("127.0.0.4")` would still keep working on for legacy devices), but this has some nice benefits:
1. You can obtain the parameters that are currently being used for a device instance and save them somewhere.
2. You can later load the saved structure to re-create an instance.

For the HA integration,  we will store the connection parameters data inside a config entry after a successful connection and just pass this dict to the library to recreate them after a reboot.

Reply:
Something I noticed yesterday is that in order to retain the "platinum" status for the integration, we need also a way to accept the session passed by homeassistant: https://developers.home-assistant.io/docs/integration_quality_scale_index/#platinum-

I think this would also fit inside the connection parameters, but necessitates changes to transport code as we must be careful not to clear the cookie jar as that's shared by all HA integrations..

Reply:
Done

Reply:
So I've added the websession parameters to this PR although not in the ConnectionParameters class as it would have complicated the serialization and de-serialization logic.  I think it's simpler this way.  To keep  changes minimal as this is quite an HA specific thing I've only added it to `discover_single()` and `connect_single()` which then just sets its on the transport.

We're using `create_async_https_client` rather than `get_async_client` because we can't share cookies with the rest of HA (the tplink devices only expect the one cookie).  I believe this is correct approach from checking on the discord and the docstring [here](https://github.com/home-assistant/core/blob/dev/homeassistant/helpers/httpx_client.py#L70C1-L70C1).